docs: update provisioner docs to contain information about registry+v1 and aggregate concepts (#424)
operator-framework/rukpak: internal/provisioner/README.md

# Provisioners

The following subdirectories contain concrete Provisioner implementations.

- [plain](plain/README.md) - provisions `plain+v0` k8s bundles
- [registry](registry/README.md) - provisions `registry+v1` OLM bundles

## Global Provisioner Concepts
Below is an example of the provisioner reconciliation flow:

```mermaid
graph TD
    C[Provisioner]
    C -->|Watches| D[Bundle]
    C -->|Watches| E[BundleInstance]
    D -->|References| F[Content]
    E -->|Creates/Manages| F[Content]
```

> ### Make bundle content available but do not install it

Reviewer: I think starting with this section is a bit confusing to users -- is inspecting bundle content without applying it the first provisioner concept we want to show off? The plain doc had the use case of installing a Bundle via a BI as the primary use case, I think it's more intuitive.

Reviewer: That section looks really good. I would argue that we should move `Make bundle content available but do not install it` down even further and move `Pivoting between bundle versions` after the new section, since that concept is more important. This was just an oversight in the original doc I wrote.

Author: Gotcha, I swapped those two now!

There is a natural separation between sourcing of the content and application of that content via two separate RukPak
APIs, `Bundle` and `BundleInstance`. A user can specify a particular `Bundle` to be available in the cluster for

> inspection before any application of the resources. Given a `Bundle` resource named `my-bundle`, the plain provisioner

Reviewer: nit: since this is the top-level provisioner README, should we have references to the plain provisioner specifically here? I would expect we could treat this doc as more generic, and defer to the individual provisioner README docs for specifics.

Author: The reference to plain here should definitely not be there. Fixing all of the reference to it in this top-level document.

> inspection before any application of the resources. Given a `Bundle` resource named `my-bundle`, the plain provisioner

Reviewer 2: ```suggestion
inspection before any application of the resources. Given a `Bundle` resource named `my-bundle`, a provisioner
```

> will pull down and unpack the bundle to a tar.gz file that is saved into a bundle cache directory mounted in the
> provisioner pods.

Reviewer 2: Bundle storage is an implementation detail of the each provisioner. At this level, it might be better to say:
```suggestion
will pull down and unpack the bundle to a tar.gz file that is saved using the provisioner's internal storage mechanism.
```

> By default, `rukpak-system` is the configured namespace for deploying `plain` provisioner-related system resources.

Reviewer 2: ```suggestion
By default, RukPak's built-in provisioners and other core resources are deployed into the `rukpak-system` namespace.
```

the following script:

```bash
BUNDLE_NAME=my-bundle

kubectl create sa fetch-bundle -n default
kubectl create clusterrolebinding fetch-bundle --clusterrole=bundle-reader --serviceaccount=default:fetch-bundle
export TOKEN=$(kubectl get secret -n default $(kubectl get sa -n default fetch-bundle -o jsonpath='{.secrets[0].name}') -o jsonpath='{.data.token}' | base64 -d)
export URL=$(kubectl get bundle $BUNDLE_NAME -o jsonpath='{.status.contentURL}')
kubectl run -qit --rm -n default --restart=Never fetch-bundle --image=curlimages/curl --overrides='{ "spec": { "serviceAccount": "fetch-bundle" }  }' --command -- curl -sSLk -H "Authorization: Bearer $TOKEN" -o - $URL | tar ztv
kubectl delete clusterrolebinding fetch-bundle
kubectl delete sa fetch-bundle -n default
```

Simplifying the process of fetching this bundle content (e.g. via a plugin) is on the RukPak roadmap.

### Pivoting between bundle versions

The `BundleInstance` API is meant to indicate the version of the bundle that should be active within the cluster.

Given an existing BundleInstance resource in the cluster, which contains an embedded Bundle template for the

> `my-bundle-v0.0.1` bundle, you can modify the desired specification and the plain provisioner will automatically generate

Reviewer 2: ```suggestion
`my-bundle-v0.0.1` bundle, you can modify the desired specification and a provisioner will automatically generate
```

a new `my-bundle-v0.0.2` Bundle matching that template.

When the new Bundle resource has been rolled out successfully, the old `my-bundle-v0.0.1` Bundle will be deleted from the cluster.

> The provisioner also continually reconciles the created content via dynamic watches to ensure that all
> resources referenced by the bundle are present on the cluster.

Reviewer 2: ```suggestion
Provisioners also continually reconcile the created content via dynamic watches to ensure that all
resources referenced by the bundle are present on the cluster.
```
